### PR TITLE
Move bunch of analyzer base types and test utility types into shared …

### PR DESCRIPTION
**Pull request by mavasani**
…layer

### CODE REVIEW

File: Compilers.sln [hunk 1]
```diff
@@ -162,26 +160,46 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Net.Compilers.Too
 EndProject
 Project("{778DAE3C-4631-46EA-AA77-85C1314464D9}") = "vbi", "src\Interactive\vbi\vbi.vbproj", "{706CFC25-B6E0-4DAA-BCC4-F6FAAFEEDF87}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.CodeAnalysis.CodeStyle.UnitTestUtilities", "src\CodeStyle\Core\Tests\Microsoft.CodeAnalysis.CodeStyle.UnitTestUtilities.csproj", "{21A59DB8-E9E0-42EB-A599-457C5BC87FC3}"
```
Comment (mavasani) [inline Compilers.sln]: Renaming `Microsoft.CodeAnalysis.CodeStyle.UnitTests` to `Microsoft.CodeAnalysis.CodeStyle.UnitTestUtilities` is the core change to the sln file and the only compiler layer change. Other changes got pulled in by VS automatically.

File: src/Analyzers/Core/Analyzers/AbstractBuiltInCodeStyleDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -8,6 +8,10 @@
 using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.Options;
 
+#if CODE_STYLE
```
Comment (mavasani) [inline src/Analyzers/Core/Analyzers/AbstractBuiltInCodeStyleDiagnosticAnalyzer.cs]: All the `#if CODE_STYLE` added in this PR are temporary until Sam's changes unifies options infrastructure in shared layer.

File: src/Analyzers/Core/Analyzers/Helpers/DiagnosticHelper.cs [hunk 3]
```diff
@@ -74,7 +73,7 @@ internal static class DiagnosticHelper
         /// </param>
         /// <param name="tagIndices">
         /// a map of location tag to index in additional locations.
-        /// <see cref="AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer{TLanguageKindEnum, TParenthesizedExpressionSyntax}"/> for an example of usage.
+        /// "AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer" for an example of usage.
```
Comment (mavasani) [inline src/Analyzers/Core/Analyzers/Helpers/DiagnosticHelper.cs]: We can revert this change once this analyzer is moved to the shared analyzers layer.

File: src/CodeStyle/CSharp/Tests/Microsoft.CodeAnalysis.CSharp.CodeStyle.UnitTests.csproj [hunk 4]
```diff
@@ -13,9 +14,19 @@
     <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="$(MicrosoftCodeAnalysisCSharpWorkspacesFixedVersion)" />
     <PackageReference Include="Microsoft.CodeAnalysis.CSharp.CodeFix.Testing.XUnit" Version="$(MicrosoftCodeAnalysisCSharpCodeFixTestingXUnitVersion)" />
   </ItemGroup>
+  <ItemGroup Label="Project References">
+    <!-- Directly reference the Workspaces project so we always test against the latest Roslyn bits -->
```
Comment (mavasani) [inline src/CodeStyle/CSharp/Tests/Microsoft.CodeAnalysis.CSharp.CodeStyle.UnitTests.csproj]: This change is needed to ensure that we always run Code style tests against the latest Roslyn Compiler and Workspaces bits.

**Comment by mavasani** [comment]
@CyrusNajmabadi Seems like your last PR merge causes a build break (https://github.com/dotnet/roslyn/pull/41416). All the subsequent PRs are failing with the following nullability error:

```src\Features\Core\Portable\UseNullPropagation\AbstractUseNullPropagationCodeFixProvider.cs#L92
src\Features\Core\Portable\UseNullPropagation\AbstractUseNullPropagationCodeFixProvider.cs(92,130): error CS8604: Possible null reference argument for parameter 'semanticModel' in 'SyntaxNode? AbstractUseNullPropagationDiagnosticAnalyzer<TSyntaxKind, TExpressionSyntax, TConditionalExpressionSyntax, TBinaryExpressionSyntax, TInvocationExpression, TMemberAccessExpression, TConditionalAccessExpression, TElementAccessExpression>.GetWhenPartMatch(ISyntaxFactsService syntaxFacts, ISemanticFactsService semanticFacts, SemanticModel semanticModel, SyntaxNode expressionToMatch, SyntaxNode whenPart)'.
```

File: src/CodeStyle/Core/CodeFixes/Microsoft.CodeAnalysis.CodeStyle.Fixes.csproj [hunk 5]
```diff
@@ -16,8 +16,8 @@
   <ItemGroup>
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.CSharp.CodeStyle.Fixes" />
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.VisualBasic.CodeStyle.Fixes" />
-    <InternalsVisibleTo Include="Microsoft.CodeAnalysis.CodeStyle.UnitTests" />
-    <InternalsVisibleTo Include="Microsoft.CodeAnalysis.CSharp.CodeStyle.UnitTests" />
+    <InternalsVisibleTo Include="Microsoft.CodeAnalysis.CodeStyle.UnitTestUtilities" />
+    <InternalsVisibleTo Include="Microsoft.CodeAnalysis.CSharp.CodeStyle.UnitTestUtilities" />
```
Comment (sharwell) [inline src/CodeStyle/Core/CodeFixes/Microsoft.CodeAnalysis.CodeStyle.Fixes.csproj]: 😕 This one shouldn't have changed names right?

File: src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier`2+Test.cs [hunk 6]
```diff
@@ -76,17 +78,25 @@ public Test()
...
             {
-                return new WorkspaceAnalyzerOptions(base.GetAnalyzerOptions(project), project.Solution);
+                var analyzerOptions = base.GetAnalyzerOptions(project);
+
+#if !CODE_STYLE
+                analyzerOptions = new WorkspaceAnalyzerOptions(analyzerOptions, project.Solution);
+#endif
+
+                return analyzerOptions;
             }
```
Comment (sharwell) [inline src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier`2+Test.cs]: 💡 Just wrap the entire method. The inherited implementation works for the code style layer.

File: src/CodeStyle/Core/CodeFixes/Microsoft.CodeAnalysis.CodeStyle.Fixes.csproj [hunk 7]
```diff
@@ -16,8 +16,8 @@
   <ItemGroup>
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.CSharp.CodeStyle.Fixes" />
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.VisualBasic.CodeStyle.Fixes" />
-    <InternalsVisibleTo Include="Microsoft.CodeAnalysis.CodeStyle.UnitTests" />
-    <InternalsVisibleTo Include="Microsoft.CodeAnalysis.CSharp.CodeStyle.UnitTests" />
+    <InternalsVisibleTo Include="Microsoft.CodeAnalysis.CodeStyle.UnitTestUtilities" />
```
Comment (sharwell) [inline src/CodeStyle/Core/CodeFixes/Microsoft.CodeAnalysis.CodeStyle.Fixes.csproj]: ❔ Why rename this assembly? We can still reference it as a test utilities library even if it's also used for authoring language-independent tests.

Comment (mavasani) [reply src/CodeStyle/Core/CodeFixes/Microsoft.CodeAnalysis.CodeStyle.Fixes.csproj]: This causes our Determinism queue on CI to fail - it checks that projects ending with `UnitTests` are not referenced by any other projects that contain unit tests.

File: src/CodeStyle/Core/CodeFixes/Microsoft.CodeAnalysis.CodeStyle.Fixes.csproj (hunk 5 above)
Comment (mavasani) [reply src/CodeStyle/Core/CodeFixes/Microsoft.CodeAnalysis.CodeStyle.Fixes.csproj]: Ah yes, let me revert.

File: src/CodeStyle/Core/CodeFixes/Microsoft.CodeAnalysis.CodeStyle.Fixes.csproj (hunk 7 above)
Comment (mavasani) [reply src/CodeStyle/Core/CodeFixes/Microsoft.CodeAnalysis.CodeStyle.Fixes.csproj]: See https://github.com/dotnet/roslyn/blob/e051790e791e0ca6ed36404db17307bea868306e/src/Tools/BuildBoss/ProjectData.cs#L17-L18

https://github.com/dotnet/roslyn/blob/bc07049760d044969409de0758aa22b3d2a48258/src/Tools/BuildBoss/ProjectCheckerUtil.cs#L214-L230

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi Seems like your last PR merge causes a build break (#41416). All the subsequent PRs are failing with the following nullability error in AbstractUseNullPropagationCodeFixProvider:

I don't understand how that's possible.  The PR was totally passing green.  Does our CI system allow merging in without revalidating the PR against master? Seems like a big hole if so.

**Comment by mavasani** [comment]
@CyrusNajmabadi Not sure. However, https://github.com/dotnet/roslyn/pull/42001 should fix the break.
